### PR TITLE
Include function execution results in the transaction results

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Include function execution results in the transaction results
- Change SET operation's result structure: `[]` -> `{ result_list: [] }`
- Always return gas_amount_total and gas_cost_total

Related issue: https://github.com/ainblockchain/ain-blockchain/issues/265

Background:
- This is needed to implement atomic native function triggering
- This is helpful for development and debugging as well

Examples:
- Transfer tx results BEFORE:
```
{
    "code": 0,
    "result": {
        "tx_hash": "0xafeefd16e1332467babe583ba8375f4389ea7f599732cba7fda99cf66920b1e5",
        "result": {
            "code": 0,
            "gas": {
                "gas_amount": {
                    "service": 1003
                }
            }
        }
    }
}
```

- Transfer tx results AFTER:
```
{
    "code": 0,
    "result": {
        "tx_hash": "0x4ddcd931d10e939d44304cc91cde5b09bfdce999b169ab6d958ed9438b6f0c5b",
        "result": {
            ".func_results": {
                "_transfer": {
                    ".op_results": [
                        {
                            "path": "/accounts/0x09A0d53FDf1c36A131938eb379b98910e55EEfe1/balance",
                            "result": {
                                "code": 0,
                                "gas_amount": 1
                            }
                        },
                        {
                            "path": "/accounts/0x7BE37246A6102270fC7dbdC2Fc199bf187500bC3/balance",
                            "result": {
                                "code": 0,
                                "gas_amount": 1
                            }
                        },
                        {
                            "path": "/transfer/0x09A0d53FDf1c36A131938eb379b98910e55EEfe1/0x7BE37246A6102270fC7dbdC2Fc199bf187500bC3/1/result",
                            "result": {
                                "code": 0,
                                "gas_amount": 1
                            }
                        }
                    ],
                    "code": "SUCCESS",
                    "gas_amount": 1000,
                }
            },
            "code": 0,
            "gas_amount": 1,
            "gas_amount_total": 1004,
            "gas_cost_total": 0
        }
    }
}
```

### CODE REVIEW

File: db/functions.js [hunk 1]
```diff
@@ -514,48 +551,46 @@ class Functions {
...
         PredefinedDbPaths.GAS_FEE, PredefinedDbPaths.GAS_FEE, blockNumber);
-    const result = this.setServiceAccountTransferOrLog(
-        from, gasFeeServiceAccountName, value.amount, auth, timestamp, transaction);
-    if (ChainUtil.isFailedTx(result)) {
-      logger.error(`  =>> gasFee collect failed: ${JSON.stringify(result)}`);
+    const result =
+        this.setServiceAccountTransferOrLog(from, gasFeeServiceAccountName, value.amount, context);
+    if (!ChainUtil.isFailedTx(result)) {
+      logger.error(`  ===> _collectFee failed: ${JSON.stringify(result)}`);
       // TODO(lia): return error, check in setValue(), revert changes
```
Comment (liayoo) [inline db/functions.js:571]: could you move this todo to inside the else { ... } ?

File: db/functions.js [hunk 2]
```diff
@@ -514,48 +551,46 @@ class Functions {
...
     const gasFeeServiceAccountName = ChainUtil.toServiceAccountName(
         PredefinedDbPaths.GAS_FEE, PredefinedDbPaths.GAS_FEE, blockNumber);
-    const result = this.setServiceAccountTransferOrLog(
-        from, gasFeeServiceAccountName, value.amount, auth, timestamp, transaction);
-    if (ChainUtil.isFailedTx(result)) {
-      logger.error(`  =>> gasFee collect failed: ${JSON.stringify(result)}`);
+    const result =
+        this.setServiceAccountTransferOrLog(from, gasFeeServiceAccountName, value.amount, context);
+    if (!ChainUtil.isFailedTx(result)) {
+      logger.error(`  ===> _collectFee failed: ${JSON.stringify(result)}`);
```
Comment (liayoo) [inline db/functions.js:570]: This should also be moved to the else?

File: db/functions.js [hunk 3]
```diff
@@ -514,48 +551,46 @@ class Functions {
...
     }
     const gasFeeServiceAccountName = ChainUtil.toServiceAccountName(
         PredefinedDbPaths.GAS_FEE, PredefinedDbPaths.GAS_FEE, blockNumber);
     const result = this.setServiceAccountTransferOrLog(
-        gasFeeServiceAccountName, proposer, gasCostTotal, auth, timestamp, transaction);
-    if (ChainUtil.isFailedTx(result)) {
-      logger.error(`  =>> gasFee distribute failed: ${JSON.stringify(result)}`);
+        gasFeeServiceAccountName, proposer, gasCostTotal, context);
+    if (!ChainUtil.isFailedTx(result)) {
+      logger.error(`  ===> _distributeFee failed: ${JSON.stringify(result)}`);
```
Comment (liayoo) [inline db/functions.js:590]: ditto

File: db/functions.js (hunk 1 above)
Comment (platfowner) [reply db/functions.js]: Just fixed. Thanks!

File: db/functions.js (hunk 2 above)
Comment (platfowner) [reply db/functions.js]: Just moved. Thanks!

File: db/functions.js (hunk 3 above)
Comment (platfowner) [reply db/functions.js]: ditto

File: db/index.js [hunk 4]
```diff
@@ -828,26 +819,28 @@ class DB {
...
-          }
-        }
+      if (blockNumber > 0) {
+        result.gas_cost_total = ChainUtil.getTotalGasCost(gasPrice, result.gas_amount_total);
       }
       if (tx && auth && auth.addr && !auth.fid) {
         this.updateAccountNonceAndTimestamp(auth.addr, tx.tx_body.nonce, tx.tx_body.timestamp);
       }
     }
+    if (result.gas_cost_total > 0) {
```
Comment (liayoo) [inline db/index.js:835]: Do we collect fees for failed txs as well?

File: common/chain-util.js [hunk 5]
```diff
@@ -290,8 +293,8 @@ class ChainUtil {
     if (!result) {
       return true;
     }
-    if (Array.isArray(result)) {
-      for (const elem of result) {
+    if (result.result_list && ChainUtil.isArray(result.result_list)) {
+      for (const elem of result.result_list) {
         if (ChainUtil.isFailedTxResultCode(elem.code)) {
           return true;
         }
```
Comment (liayoo) [inline common/chain-util.js:300]: Should we check for the function results as well?

Comment (platfowner) [reply common/chain-util.js:301]: Right. Please consider this PR as a feature parity implementation. Just added a todo for such next steps.

File: db/index.js (hunk 4 above)
Comment (platfowner) [reply db/index.js]: Just reverted. However it seems to be worth considering charging gas fee for failure cases as well. Let's think about it later. Thanks.

File: common/chain-util.js (hunk 5 above)
Comment (liayoo) [reply common/chain-util.js:301]: I see, thanks for the explanation!

File: db/functions.js [hunk 6]
```diff
@@ -387,122 +370,176 @@ class Functions {
...
+    } else {
+      return this.returnFuncResult(context, FunctionResultCode.FAILURE);
+    }
+  }
 
-    return this.setValueOrLog(
-        ChainUtil.formatPath(lastTxPath), { tx_hash: transaction.hash }, auth, timestamp,
-        transaction);
+  /**
+   * Erases the value.
```
Comment (liayoo) [inline db/functions.js:488]: Is this native function for testing purposes? If so, could you leave a note about that?

Comment (platfowner) [reply db/functions.js]: Done, thanks!

**Review by liayoo — APPROVED** [review]
LGTM, thanks!!

**Comment by platfowner** [comment]
Thanks for the review. Now merging..

**Pull request merged**